src/Action/LoginAction.php

```diff
@@ -0,0 +1,149 @@
+     */
+    private $urlGenerator;
+
+    /**
+     * @var AuthorizationCheckerInterface
+     */
+    private $authorizationChecker;
+
+    /**
+     * @var Pool
+     */
+    private $adminPool;
+
+    /**
+     * @var TemplateRegistryInterface
+     */
+    private $templateRegistry;
+
+    /**
+     * @var TokenStorageInterface
+     */
+    private $tokenStorage;
+
+    /**
+     * @var Session
+     */
+    private $session;
+
+    /**
+     * @var CsrfTokenManagerInterface
+     */
+    private $csrfTokenManager;
+
+    public function __construct(
+        EngineInterface $templating,
+        UrlGeneratorInterface $urlGenerator,
+        AuthorizationCheckerInterface $authorizationChecker,
+        Pool $adminPool,
+        TemplateRegistryInterface $templateRegistry,
+        TokenStorageInterface $tokenStorage,
+        Session $session
+    ) {
+        $this->templating = $templating;
+        $this->urlGenerator = $urlGenerator;
+        $this->authorizationChecker = $authorizationChecker;
+        $this->adminPool = $adminPool;
+        $this->templateRegistry = $templateRegistry;
+        $this->tokenStorage = $tokenStorage;
+        $this->session = $session;
+    }
+
+    public function __invoke(Request $request): Response
+    {
+        $token = $this->tokenStorage->getToken();
+
+        if ($token) {
+            $user = $token->getUser();
+
+            if ($user instanceof UserInterface) {
+                $this->session->getFlashBag()->add('sonata_user_error', 'sonata_user_already_authenticated');
+
+                return new RedirectResponse($this->urlGenerator->generate('sonata_admin_dashboard'));
+            }
+        }
+
+        $session = $request->getSession();
+
+        $authErrorKey = Security::AUTHENTICATION_ERROR;
+
+        // get the error if any (works with forward and redirect -- see below)
+        if ($request->attributes->has($authErrorKey)) {
+            $error = $request->attributes->get($authErrorKey);
+        } elseif (null !== $session && $session->has($authErrorKey)) {
+            $error = $session->get($authErrorKey);
+            $session->remove($authErrorKey);
+        } else {
+            $error = null;
+        }
+
+        if (!$error instanceof AuthenticationException) {
+            $error = null; // The value does not come from the security component.
+        }
+
+        if ($this->authorizationChecker->isGranted('ROLE_ADMIN')) {
+            $refererUri = $request->server->get('HTTP_REFERER');
+            $url = $refererUri && $refererUri != $request->getUri() ? $refererUri : $this->urlGenerator->generate('sonata_admin_dashboard');
+
+            return new RedirectResponse($url);
+        }
+
+        $csrfToken = null;
+        if ($this->csrfTokenManager) {
+            $csrfToken = $this->csrfTokenManager->getToken('authenticate')->getValue();
+        }
+
+        return $this->templating->renderResponse('@SonataUser/Admin/Security/login.html.twig', [
```

Comment:
It looks like the `templating` service is defined by FrameworkBundle which is a requirement. So I think we always fell to this definition in case of rendering:
https://github.com/symfony/framework-bundle/blob/master/Resources/config/templating_php.xml
https://github.com/symfony/framework-bundle/blob/4.1/DependencyInjection/FrameworkExtension.php#L823